### PR TITLE
chore: support drop_last=False by default for val dataloader

### PR DESCRIPTION
**Pull request by peri044**
# What does this PR do ?

Supports drop_last=False in validation. Partially fixes https://github.com/NVIDIA-NeMo/RL/issues/605. Please review and let me know if we need the configurable parameter (`policy.drop_last_validation`)

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

**Comment by peri044** [comment]
@terrykong  please review

**Comment by jveronvialard** [comment]
Thanks! Should `nemo_rl/algorithms/grpo.py` also be updated? Are there any use-cases where drop_last=True might be useful? If so, a configurable parameter might be better.

**Comment by peri044** [comment]
@jveronvialard @parthchadha 
I revisited this issue and tested it on a system with 2 GPUs, though the observations apply more generally to any `n_gpus > 1` setup.

> Are there any use-cases where drop_last=True might be useful?

Yes—particularly for DPO and reward model training. In these setups, each sample consists of a chosen and a rejected response, which are split into two separate entries during batch collation.

Let’s take an example: suppose there’s only 1 sample left in the final global batch. That results in a batch size of 2 (e.g., input_ids would have shape [2, seq_len]). Normally, with multiple GPUs, this batch would be sharded. However, for this last batch, I wanted to avoid sharding and instead use a single worker to process the entire batch.

I tried using `run_multiple_workers_single_data` (note: `run_single_worker_single_data` hangs in my setup). But here's the problem: inside this function, `local_gbs` becomes 1 because the total global batch size is 2 and `self.dp_size=2`
(see: https://github.com/NVIDIA-NeMo/RL/blob/main/nemo_rl/models/policy/megatron_policy_worker.py#L742).

The idea here was for the last small batch, all workers process the same data, but only one (say, `worker_idx=0`) to return results. However, since `local_gbs=1`, the preference loss function ends up receiving only one half of the pair (either the chosen or rejected response), leading to invalid input. 

To summarize:
- When `drop_last=False`, the last batch may be small and not suitable for typical sharding.
- For this last small batch, I’ve disabled sharding and routed the data to a single worker.
- However, due to how `local_gbs` is computed, we still end up with partial data—just one element from the pair (either chosen or rejected)—which breaks the preference loss function.

File: nemo_rl/models/policy/__init__.py [hunk 1]
```diff
@@ -141,6 +141,7 @@ class PolicyConfig(TypedDict):
     make_sequence_length_divisible_by: int
     max_total_sequence_length: int
     max_grad_norm: NotRequired[Union[float, int]]
+    drop_last_validation: NotRequired[bool]
```
Comment (odelalleau) [inline nemo_rl/models/policy/__init__.py]: Can we rename this `val_drop_last` to be consistent with other validation config options that all start with `val`?

File: nemo_rl/distributed/batched_data_dict.py [hunk 2]
```diff
@@ -187,8 +187,12 @@ def chunk(self, rank: int, chunks: int) -> "SlicedDataDict":
 
         return chunked_batch
 
-    def reorder_data(self, reorded_indices: list[int]):
-        """Reorders the data along the batch dimension by the given indices."""
+    def get_total_batch_size(self: Self) -> int:
+        """Get the total batch size of the batch.
```
Comment (odelalleau) [inline nemo_rl/distributed/batched_data_dict.py]: Why is this called "the total batch size"? (since it doesn't sum batch sizes, so it doesn't look like a total)

File: nemo_rl/models/policy/lm_policy.py [hunk 3]
```diff
@@ -320,7 +320,14 @@ def train(
         """Train the policy on a batch of data with a given loss function."""
         batch_size = gbs or self.cfg["train_global_batch_size"]
         micro_batch_size = mbs or self.cfg["train_micro_batch_size"]
+
         # Shard and replicate the batch
+        allow_uneven_shards = False
+        total_batch_size = data.get_total_batch_size()
+        if total_batch_size % batch_size != 0:
+            batch_size = total_batch_size
+            allow_uneven_shards = True
```
Comment (odelalleau) [inline nemo_rl/models/policy/lm_policy.py]: Can you please add a comment to explain when this happens? (I don't really understand what's going on there -- is it only when `total_batch_size < batch_size`?)

**Review by odelalleau — COMMENTED** [review]
Couple of questions below, also regarding:

> we still end up with partial data—just one element from the pair (either chosen or rejected)—which breaks the preference loss function.

We'll need to find a way to fix this.

File: nemo_rl/distributed/batched_data_dict.py (hunk 2 above)
Comment (peri044) [reply nemo_rl/distributed/batched_data_dict.py]: fixed it now

File: nemo_rl/models/policy/lm_policy.py (hunk 3 above)
Comment (peri044) [reply nemo_rl/models/policy/lm_policy.py]: Thanks for the review @odelalleau. that's right `allow_uneven_shards=True` happens when last batch size is smaller than `gbs`. I added a comment and renamed other variables accordingly.

**Comment by peri044** [comment]
> Couple of questions below, also regarding:
> 
> > we still end up with partial data—just one element from the pair (either chosen or rejected)—which breaks the preference loss function.
> 
> We'll need to find a way to fix this.

@odelalleau  I think I found a way to fix this issue. Please review the latest commits. 
Here's what I'm doing
1) Disable the sharding for the last small batch.
2) Process the last small batch on all the workers (`run_all_workers_single_data`) and gather the result from worker_idx=0. In each of this worker, the `local_gbs` should be equal to size of small batch (I had to do `gbs=batch_size * dp_size` to ensure this is happening correctly)

**Comment by peri044** [comment]
We might be able to do away with this `use_preference_model` distinction and apply the same logic to SFT/GRPO as well. I'll try that and update this PR.

**Comment by peri044** [comment]
I've updated the logic so that both SFT and DPO models handle the final small batch consistently—without sharding. Please take a look. We probably need a better identifier (currently using `use_preference_model`) to determine the minimum `micro_batch_size` that ensures `data.size % micro_batch_size == 0`.

File: nemo_rl/algorithms/dpo.py [hunk 4]
```diff
@@ -180,7 +180,7 @@ def setup(
                 "make_sequence_length_divisible_by"
             ],
         ),
-        drop_last=True,
+        drop_last=policy_config.get("val_drop_last", True),
```
Comment (jveronvialard) [inline nemo_rl/algorithms/dpo.py]: Could you also update the examples/configs and docs/guides?

File: nemo_rl/models/policy/lm_policy.py [hunk 5]
```diff
@@ -316,60 +320,90 @@ def train(
...
-                batch_size=batch_size,
-                sequence_packing_args=self.sequence_packing_args,
+            object_refs = self.worker_group.run_all_workers_single_data(
+                "train",
+                data=data,
+                loss_fn=loss_fn,
+                eval_mode=eval_mode,
+                # Note: On each worker, local_gbs = gbs / dp_size.
+                # To handle the small batch properly, we scale its size by dp_size so each worker receives exactly one local batch.
+                # We set mbs = 1 (or 2 for preference model) to ensure that the small batch size is divisible by the micro_batch_size.
```
Comment (jveronvialard) [inline nemo_rl/models/policy/lm_policy.py:376]: Should you assert this? E.g. in case it's a preference model and the batch_size is an odd number

File: nemo_rl/models/policy/lm_policy.py [hunk 6]
```diff
@@ -316,60 +320,90 @@ def train(
         eval_mode: bool = False,
         gbs: Optional[int] = None,
         mbs: Optional[int] = None,
+        is_preference_model: bool = False,
```
Comment (jveronvialard) [inline nemo_rl/models/policy/lm_policy.py]: How about `mbs_last_val: int = 1` instead? And set it to 2 in dpo.py and rm.py

File: nemo_rl/models/policy/lm_policy.py [hunk 7]
```diff
@@ -316,60 +320,90 @@ def train(
...
-            self.dynamic_batching_args["max_tokens_per_microbatch"] = self.cfg[
-                "dynamic_batching"
-            ]["train_mb_tokens"]
-            sharded_data, _ = data.shard_by_batch_size(
-                dp_size,
-                batch_size=batch_size,
-                dynamic_batching_args=self.dynamic_batching_args,
+
+        # When drop_last_val = False in the dataloader, the last batch will be smaller than batch_size.
+        if data.size < batch_size:
```
Comment (jveronvialard) [inline nemo_rl/models/policy/lm_policy.py:353]: Should you assert that this happens only for the last validation batch?

File: nemo_rl/models/policy/lm_policy.py [hunk 8]
```diff
@@ -316,60 +320,90 @@ def train(
...
-        if self.use_dynamic_batches:
-            self.dynamic_batching_args["max_tokens_per_microbatch"] = self.cfg[
-                "dynamic_batching"
-            ]["train_mb_tokens"]
-            sharded_data, _ = data.shard_by_batch_size(
-                dp_size,
-                batch_size=batch_size,
-                dynamic_batching_args=self.dynamic_batching_args,
+
+        # When drop_last_val = False in the dataloader, the last batch will be smaller than batch_size.
```
Comment (jveronvialard) [inline nemo_rl/models/policy/lm_policy.py]:
```suggestion
        # When drop_last_val = False in the dataloader, the last batch can be smaller than batch_size.
```

File: nemo_rl/algorithms/rm.py [hunk 9]
```diff
@@ -248,7 +248,11 @@ def validate(
         num_valid_batches = 0
 
         policy.prepare_for_training()
+
         for batch_idx, val_batch in enumerate(val_dataloader):
+            print(
```
Comment (jveronvialard) [inline nemo_rl/algorithms/rm.py]: Should this print statement be removed?

File: nemo_rl/algorithms/dpo.py (hunk 4 above)
Comment (jveronvialard) [reply nemo_rl/algorithms/dpo.py]: Also, this new feature is a bit hidden. Instead, should all configs be updated and then use `policy_config["val_drop_Last"]`?

**Review by jveronvialard — CHANGES_REQUESTED** [review]
Thanks for the PR! 
Could you add convergence plots comparing drop_last_batch=True and drop_last_batch=False? If possible, using a small validation dataset (you could truncate the existing HelpSteer3 one), so the effect is observable. Could you also add unit tests?

**Comment by ashors1** [comment]
I haven't finished reviewing, but I wanted to share that I was also working on a more naive approach that just pads the final batch to the global batch size: https://github.com/NVIDIA-NeMo/RL/compare/main...ashors/drop-last-false. Thought I would share here in case anyone has any comments on my more naive approach vs this one

**Comment by ashors1** [comment]
> I haven't finished reviewing, but I wanted to share that I was also working on a more naive approach that just pads the final batch to the global batch size: [main...ashors/drop-last-false](https://github.com/NVIDIA-NeMo/RL/compare/main...ashors/drop-last-false). Thought I would share here in case anyone has any comments on my more naive approach vs this one

After taking a closer look a this PR, I think I am slightly partial to the more naive approach of just padding the partial batch because it decouples the drop_last logic from the policies. But interested to hear other's thoughts @peri044 @odelalleau @jveronvialard

**Comment by peri044** [comment]
> > I haven't finished reviewing, but I wanted to share that I was also working on a more naive approach that just pads the final batch to the global batch size: [main...ashors/drop-last-false](https://github.com/NVIDIA-NeMo/RL/compare/main...ashors/drop-last-false). Thought I would share here in case anyone has any comments on my more naive approach vs this one
> 
> After taking a closer look a this PR, I think I am slightly partial to the more naive approach of just padding the partial batch because it decouples the drop_last logic from the policies. But interested to hear other's thoughts @peri044 @odelalleau @jveronvialard

@ashors1  I think this works well too. Here are my thoughts/concerns in order of priority
1) validate() function is implemented at a bunch of places (grpo.py, sft.py, rm.py, dpo.py, eval.py and other algos). so the padding approach would exist at multiple places which might be high maintenance cost.
2) Padding approach might change based on preference dataset (DPO has 2 responses but some other algos might have 3 responses ?). In this PR case, this issue would exist as well and the solution would be setting mbs value = 3.
3) If batch_size is high, will padding be expensive ? This is actually contradicting my recent validation performance experiments though

 > it separates the drop_last=False logic from the policies
 
I see. I think my intention was to generally support small (non-divisible) batches using the worker utilities we have. This should be applicable to training and validation although I agree training is not important.  We could go with your approach if you think we shouldn't be touching policy/workers to support this.

File: nemo_rl/models/policy/lm_policy.py (hunk 7 above)
Comment (peri044) [reply nemo_rl/models/policy/lm_policy.py:353]: Any insights how we detect if this `data` is indeed the last validation batch ? I also wonder if we should make this assertion here if it's needed.

File: nemo_rl/models/policy/lm_policy.py (hunk 5 above)
Comment (peri044) [reply nemo_rl/models/policy/lm_policy.py:376]: Added an assert now. PTAL.

File: nemo_rl/algorithms/dpo.py (hunk 4 above)
Comment (peri044) [reply nemo_rl/algorithms/dpo.py]: updated all the configs now and used policy_config["val_drop_last"] instead

**Comment by jveronvialard** [comment]
Since this happens only for the last batch, I think I also have a slight preference for @ashors1 's padding approach. But to @peri044 's point, this padding should be easy extended to other algorithms. How about implementing it at the dataset `AllTaskProcessedDataset` level? On a side note, @peri044 , in your PR you might need to check how the metrics are being normalized too.

@peri044 , a classification head is initialized at each run, so you might observe such differences between runs (at start, it should reduce after a few training steps).

**Comment by ashors1** [comment]
Closing in favor of https://github.com/NVIDIA-NeMo/RL/pull/1029
